File: Editorial/0148_Sort_List/C/1_Top_Down_Merge_Sort.c
struct ListNode* getMid(struct ListNode* head);
struct ListNode* merge(struct ListNode* list1, struct ListNode* list2);

struct ListNode* sortList(struct ListNode* head) {
    if (head == NULL || head->next == NULL) return head;
    struct ListNode* mid = getMid(head);
    struct ListNode* left = sortList(head);
    struct ListNode* right = sortList(mid);
    return merge(left, right);
}

struct ListNode* merge(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode dummyHead;
    struct ListNode* ptr = &dummyHead;
    while (list1 && list2) {
        if (list1->val < list2->val) {
            ptr->next = list1;
            list1 = list1->next;
        } else {
            ptr->next = list2;
            list2 = list2->next;
        }
        ptr = ptr->next;
    }
    ptr->next = list1 ? list1 : list2;
    return dummyHead.next;
}

struct ListNode* getMid(struct ListNode* head) {
    struct ListNode* midPrev = NULL;
    while (head && head->next) {
        midPrev = (midPrev == NULL) ? head : midPrev->next;
        head = head->next->next;
    }
    struct ListNode* mid = midPrev->next;
    midPrev->next = NULL;
    return mid;
}

File: Editorial/0101-0200/0148_Sort_List/TypeScript/1_Top_Down_Merge_Sort.ts
function sortList(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head;

    let mid = getMid(head);
    let left = sortList(head);
    let right = sortList(mid);
    return merge(left, right);

    function merge(
        list1: ListNode | null,
        list2: ListNode | null,
    ): ListNode | null {
        let dummyHead = new ListNode(0, null);
        let tail = dummyHead;
        while (list1 && list2) {
            if (list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }
        tail.next = list1 ? list1 : list2;
        return dummyHead.next;
    }

    function getMid(head: ListNode | null): ListNode | null {
        let midPrev = null;
        while (head && head.next) {
            midPrev = midPrev === null ? head : midPrev.next;
            head = head.next.next;
        }
        let mid = midPrev.next;
        midPrev.next = null;
        return mid;
    }
}
